### DISCUSSION TITLE
Collision detection for "solid" shapes

### DISCUSSION THREAD

**Discussion by Toshik** [flame-engine/flame#1847] [Q&A] [answered]
Hi there.

I am trying to detect if one simple shape (circle) intersects with another shape (square). I am using CollisionCallbacks.
But here I have an issue: if circle passes over bigger one square I get onCollisionStart then onCollisionEnd when circle is actually inside the square. So, I guess onCollisionStart being called when shape boundaries intersects, so shapes are "hollow".

Is there a way to detect collisions like shapes are solid?

Sample picture captured from "Multiple Shapes" example. Blue circle does not collide with green square, and green circle collides with the same square.
![image](https://user-images.githubusercontent.com/170579/184900945-c8baf48e-df26-4e37-98d4-9ba6c772a166.png)

Thanks for any help.

**Comment by spydon** [comment, marked as answer]
The hitboxes are indeed hollow, and we don't have any out of the box support for checking those type of collisions.
What you can do is to check if the `aabb` of the hitbox is completely contained within the other hitbox's `aabb` with `containsAabb2`, this will only give a rough estimation of whether there is a collision though, but it would work for most cases.
https://api.flutter.dev/flutter/vector_math/Aabb2/containsAabb2.html

**Reply by Toshik** [reply]
Thank you for the clue, it is exactly what I need!
So, inside `onCollisionEnd` I am just checking if shape is still inside or not.

**Reply by spydon** [reply]
You might have to check `intersectsAabb2` too btw, if you don't get the desired results with only the contains function.

**Reply by Toshik** [reply]
As far as I understand, `intersectsAabb2` here should work the same way as `onCollisionStart`.
So, if I've got `onCollisionStart` - then `intersectsAabb2` gives me `true` for the same pair of hitboxes?

**Reply by spydon** [reply]
The `aabb2` is an axis aligned bounding box, so it covers a bit more area than the hitboxes themselves, but I realize that you will probably get many false positives if using the intersect method to see whether it is within the other hitbox, since it can just as well return true when it has just moved out of the other hitbox. You could also check if the center point of the hitbox is contained within the other hitbox in `onCollisionEnd`, if the `containsAabb2` doesn't give the expected result.

**Reply by st-pasha** [reply]
Note that once we implement hitboxes based on the `Shape` class (#1835), the intersection algorithm there already treats shapes as solid.

**Reply by spydon** [reply]
Hmm, I think we'll have to be able support both @st-pasha, otherwise hitboxes like the `ScreenHitbox` won't work.

**Reply by st-pasha** [reply]
ScreenHitbox is an interesting case. We have several possibilities here: (1) declare it as non-convex and intersect using the fallback wireframe mechanism, (2) make it a union of 4 rectangles that cover the outside of the screen, (3) leave it as a simple rectangle, but treat the "collisionStart" event as meaning the object is visible on screen, and "collisionEnd" meaning as it left the screen.

**Reply by spydon** [reply]
It's not only for the `ScreenHitbox`, things like this should also be possible to do for example:
https://twitter.com/wolfenrain/status/1551873822323888131
